File: ios/SiezaGym/Services/AuthService.swift
import FirebaseAuth
import FirebaseCore
import Foundation
import GoogleSignIn
import UIKit
import Observation
import os

private nonisolated let log = Logger(subsystem: "com.siezagym.app", category: "auth")

/// Sesion del usuario. Envuelve FirebaseAuth y expone solo lo que la UI necesita.
@Observable
final class AuthService {
    enum State: Equatable {
        /// Todavia no sabemos si hay sesion guardada: no hay que mostrar el login.
        case loading
        case signedOut
        case signedIn(uid: String, email: String?)
    }

    private(set) var state: State = .loading
    private(set) var errorMessage: String?
    private(set) var isWorking = false

    // Se escribe una sola vez en `init` y se lee una sola vez en `deinit`,
    // cuando ya no queda ninguna otra referencia viva. No hay carrera posible,
    // pero `deinit` es nonisolated y no puede tocar estado del main actor.
    private nonisolated(unsafe) var listener: AuthStateDidChangeListenerHandle?

    var uid: String? {
        if case let .signedIn(uid, _) = state { return uid }
        return nil
    }

    init() {
        listener = Auth.auth().addStateDidChangeListener { [weak self] _, user in
            // El listener ya llega en el main thread, pero el SDK no lo declara,
            // asi que se afirma en vez de saltar de hilo y perder el frame.
            MainActor.assumeIsolated {
                guard let self else { return }
                self.state = user.map { .signedIn(uid: $0.uid, email: $0.email) } ?? .signedOut
            }
        }
    }

    deinit {
        if let listener { Auth.auth().removeStateDidChangeListener(listener) }
    }

    func signIn(email: String, password: String) async {
        await run { try await Auth.auth().signIn(withEmail: email, password: password) }
    }

    func signUp(email: String, password: String, displayName: String) async {
        await run {
            let result = try await Auth.auth().createUser(withEmail: email, password: password)
            let name = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
            if !name.isEmpty {
                let request = result.user.createProfileChangeRequest()
                request.displayName = name
                try await request.commitChanges()
            }
            try await GymRepository().ensureProfile(
                uid: result.user.uid,
                email: result.user.email,
                displayName: name.isEmpty ? nil : name,
                provider: "password"
            )
        }
    }

    /// Login con Google. Termina en la misma cuenta de Firebase que usa la web:
    /// si entraste con Google en `sieza-gym.vercel.app`, es el mismo uid.
    func signInWithGoogle() async {
        await run {
            guard let clientID = FirebaseApp.app()?.options.clientID else {
                throw SignInError.missingClientID
            }
            guard let presenter = Self.topViewController() else {
                throw SignInError.noPresenter
            }

            GIDSignIn.sharedInstance.configuration = GIDConfiguration(clientID: clientID)
            let result = try await GIDSignIn.sharedInstance.signIn(withPresenting: presenter)

            guard let idToken = result.user.idToken?.tokenString else {
                throw SignInError.missingIDToken
            }
            let credential = GoogleAuthProvider.credential(
                withIDToken: idToken,
                accessToken: result.user.accessToken.tokenString
            )
            let authResult = try await Auth.auth().signIn(with: credential)

            // La web crea el perfil del lado del servidor al iniciar sesion; en
            // iOS no hay servidor, asi que lo hace la app con los mismos campos.
            let user = authResult.user
            try await GymRepository().ensureProfile(
                uid: user.uid,
                email: user.email,
                displayName: user.displayName,
                photoURL: user.photoURL?.absoluteString,
                provider: "google.com"
            )
        }
    }

    /// El SDK de Google necesita un UIViewController desde el que presentarse, y
    /// SwiftUI no expone ninguno.
    private static func topViewController() -> UIViewController? {
        let scene = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
        var top = scene?.keyWindow?.rootViewController
        while let presented = top?.presentedViewController {
            top = presented
        }
        return top
    }

    enum SignInError: LocalizedError {
        case missingClientID
        case noPresenter
        case missingIDToken

        var errorDescription: String? {
            switch self {
            case .missingClientID:
                "Falta el CLIENT_ID en GoogleService-Info.plist."
            case .noPresenter:
                "No se pudo abrir la ventana de Google."
            case .missingIDToken:
                "Google no devolvió el token de la cuenta."
            }
        }
    }

    func signOut() {
        do {
            // Sin esto Google recuerda la cuenta y el proximo login entra solo,
            // sin dejar elegir otra.
            GIDSignIn.sharedInstance.signOut()
            try Auth.auth().signOut()
            errorMessage = nil
        } catch {
            log.error("signOut fallo: \(error.localizedDescription, privacy: .public)")
            errorMessage = "No se pudo cerrar la sesión."
        }
    }

    func clearError() { errorMessage = nil }

    private func run(_ operation: () async throws -> Void) async {
        isWorking = true
        errorMessage = nil
        defer { isWorking = false }
        do {
            try await operation()
        } catch let error as NSError
            where error.domain == kGIDSignInErrorDomain
            && error.code == GIDSignInError.canceled.rawValue {
            // El usuario cerro la ventana de Google a proposito.
            return
        } catch {
            log.error("auth fallo: \(error.localizedDescription, privacy: .public)")
            errorMessage = Self.readableMessage(for: error)
        }
    }

    /// Los mensajes de FirebaseAuth vienen en ingles y son tecnicos.
    static func readableMessage(for error: Error) -> String {
        if let signInError = error as? SignInError {
            return signInError.localizedDescription
        }
        guard (error as NSError).domain == AuthErrorDomain,
              let code = AuthErrorCode(rawValue: (error as NSError).code) else {
            return "Algo salió mal. Probá de nuevo."
        }
        return switch code {
        case .invalidEmail: "Ese email no es válido."
        case .emailAlreadyInUse: "Ya hay una cuenta con ese email."
        case .weakPassword: "La contraseña necesita al menos 6 caracteres."
        case .wrongPassword, .invalidCredential: "Email o contraseña incorrectos."
        case .userNotFound: "No encontramos una cuenta con ese email."
        case .networkError: "Sin conexión. Revisá internet."
        case .tooManyRequests: "Demasiados intentos. Esperá un momento."
        case .accountExistsWithDifferentCredential:
            "Ya tenés una cuenta con ese email creada de otra forma."
        default: "Algo salió mal. Probá de nuevo."
        }
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/Services/AuthService.kt
package com.siezagym.app.Services

import android.content.Context
import android.content.Intent
import android.util.Log
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.firebase.auth.AuthErrorCodes
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthException
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.UserProfileChangeRequest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

/** Sesión del usuario. Envuelve FirebaseAuth y expone solo lo que la UI necesita. */
class AuthService(private val appContext: Context) {
    private val auth = FirebaseAuth.getInstance()
    private val repository = GymRepository()

    sealed interface State {
        data object Loading : State
        data object SignedOut : State
        data class SignedIn(val uid: String, val email: String?) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    private val _errorMessage = MutableStateFlow<String?>(null)
    private val _isWorking = MutableStateFlow(false)

    val state: StateFlow<State> = _state.asStateFlow()
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    val isWorking: StateFlow<Boolean> = _isWorking.asStateFlow()

    init {
        auth.addAuthStateListener { firebaseAuth ->
            val user = firebaseAuth.currentUser
            _state.value = if (user != null) State.SignedIn(user.uid, user.email) else State.SignedOut
        }
    }

    val uid: String?
        get() = (_state.value as? State.SignedIn)?.uid

    // MARK: - Email / Contraseña

    suspend fun signIn(email: String, password: String) = attempt {
        val result = auth.signInWithEmailAndPassword(email.trim(), password).await()
        repository.ensureProfile(
            uid = result.user.uid,
            email = result.user.email,
            displayName = result.user.displayName,
            photoUrl = result.user.photoUrl?.toString(),
            provider = "password",
        )
    }

    suspend fun signUp(email: String, password: String, displayName: String) = attempt {
        val result = auth.createUserWithEmailAndPassword(email.trim(), password).await()
        val name = displayName.trim()
        if (name.isNotEmpty()) {
            result.user.updateProfile(
                UserProfileChangeRequest.Builder().setDisplayName(name).build()
            ).await()
        }
        repository.ensureProfile(
            uid = result.user.uid,
            email = result.user.email,
            displayName = name.ifEmpty { null },
            photoUrl = result.user.photoUrl?.toString(),
            provider = "password",
        )
    }

    // MARK: - Google

    /** El intent del flujo de Google. Sin `default_web_client_id` (falta la
     *  config de Firebase) devuelve null y la UI muestra cómo generarla. */
    fun googleSignInIntent(): Intent? {
        val webClientId = stringByIdentifier("default_web_client_id") ?: return null
        val client = googleClient(webClientId)
        return client.signInIntent
    }

    /** Termina el flujo de Google con el resultado que devolvió la Activity.
     *  Termina en la misma cuenta de Firebase que usa la web: si entraste con
     *  Google en sieza-gym.vercel.app, es el mismo uid. */
    suspend fun completeGoogleSignIn(data: Intent?) = attempt {
        val account: GoogleSignInAccount = try {
            GoogleSignIn.getSignedInAccountFromIntent(data).getResult(ApiException::class.java)
        } catch (e: ApiException) {
            if (e.statusCode == CommonStatusCodes.SIGN_IN_CANCELLED) throw GoogleCanceledException()
            throw e
        }

        val idToken = account.idToken
            ?: throw IllegalStateException("Google no devolvió el token de la cuenta.")
        val credential = GoogleAuthProvider.getCredential(idToken, null)
        val result = auth.signInWithCredential(credential).await()

        // La web crea el perfil del lado del servidor al iniciar sesión; en
        // Android no hay servidor, así que lo hace la app con los mismos campos.
        repository.ensureProfile(
            uid = result.user.uid,
            email = result.user.email,
            displayName = result.user.displayName,
            photoUrl = result.user.photoUrl?.toString(),
            provider = "google.com",
        )
    }

    // MARK: - Cerrar sesión

    fun signOut() {
        // Sin esto Google recuerda la cuenta y el próximo login entra solo, sin
        // dejar elegir otra. El cierre de Google corre async y no bloquea el de
        // Firebase: fallar ahí no debería impedir salir.
        webClientIdOrNull()?.let { googleClient(it).signOut().addOnFailureListener { } }
        try {
            auth.signOut()
            _errorMessage.value = null
        } catch (e: Exception) {
            Log.w(TAG, "signOut fallo: ${e.message}")
            _errorMessage.value = "No se pudo cerrar la sesión."
        }
    }

    fun clearError() { _errorMessage.value = null }

    // MARK: - Implementación

    /** El usuario cerró la ventana de Google a propósito: no es un error. */
    private class GoogleCanceledException : RuntimeException()

    private suspend fun attempt(block: suspend () -> Unit) {
        _isWorking.value = true
        _errorMessage.value = null
        try {
            block()
        } catch (e: GoogleCanceledException) {
            // Silencioso.
        } catch (e: Exception) {
            Log.w(TAG, "auth fallo: ${e.message}")
            _errorMessage.value = readableAuthError(e)
        } finally {
            _isWorking.value = false
        }
    }

    private fun stringByIdentifier(name: String): String? {
        val id = appContext.resources.getIdentifier(name, "string", appContext.packageName)
        return if (id == 0) null else appContext.getString(id)
    }

    private val webClientIdOrNull: String?
        get() = stringByIdentifier("default_web_client_id")

    private fun googleClient(webClientId: String): GoogleSignInClient {
        val options = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(webClientId)
            .requestEmail()
            .build()
        return GoogleSignIn.getClient(appContext, options)
    }

    companion object {
        private const val TAG = "AuthService"
    }
}

/** Los mensajes de FirebaseAuth vienen en inglés y son técnicos. Función pura:
 *  así los unit tests la prueban sin necesidad de una sesión real. */
internal fun readableAuthError(error: Exception): String {
    return when (error) {
        is FirebaseAuthException -> when (error.errorCode) {
            AuthErrorCodes.INVALID_EMAIL -> "Ese email no es válido."
            AuthErrorCodes.EMAIL_ALREADY_IN_USE -> "Ya hay una cuenta con ese email."
            AuthErrorCodes.WEAK_PASSWORD -> "La contraseña necesita al menos 6 caracteres."
            AuthErrorCodes.WRONG_PASSWORD,
            AuthErrorCodes.INVALID_CREDENTIAL,
            AuthErrorCodes.USER_DISABLED,
            -> "Email o contraseña incorrectos."
            AuthErrorCodes.USER_NOT_FOUND -> "No encontramos una cuenta con ese email."
            AuthErrorCodes.NETWORK_REQUEST_FAILED -> "Sin conexión. Revisá internet."
            AuthErrorCodes.TOO_MANY_ATTEMPTS_TRY_LATER -> "Demasiados intentos. Esperá un momento."
            AuthErrorCodes.ACCOUNT_EXISTS_WITH_DIFFERENT_CREDENTIAL ->
                "Ya tenés una cuenta con ese email creada de otra forma."
            else -> "Algo salió mal. Probá de nuevo."
        }
        else -> "Algo salió mal. Probá de nuevo."
    }
}